### PR TITLE
feat(vscode): add $testthat problem matcher for test failures

### LINKED ISSUE #206: testthat problem matcher
## Summary

Add a VS Code problem matcher for testthat output so that test failures appear in the Problems panel with clickable file:line links.

## What this enables

When `devtools::test()` or `testthat::test_file()` runs and a test fails, the output looks like:

```text
── Failure (test-helpers.R:12:3): process_data handles NAs ──
Expected 5, got 4.
```

A problem matcher parses this into structured diagnostics in VS Code's Problems panel, letting you click directly to the failing test line.

## Implementation

Add to `package.json` under `contributes.problemMatchers`:

```json
{
  "name": "testthat",
  "owner": "R",
  "severity": "error",
  "fileLocation": ["relative", "${workspaceFolder}/tests/testthat"],
  "pattern": [
    {
      "regexp": "^[-─ ]*(Failure|Error)\\s\\((.*\\.[Rr]):(\\d+):?(\\d+)?\\):\\s(.*)",
      "file": 2,
      "line": 3,
      "column": 4,
      "message": 5
    },
    {
      "regexp": "^(.*)$",
      "message": 1
    }
  ]
}
```

Also add a task definition of type `R` so users can define test tasks in `tasks.json` that reference `$testthat`.

## Related

- #205 (R package dev commands) — the `R Package: Test` command should use this problem matcher.

### PR DESCRIPTION
**Pull request by jbearak**
## Summary

- Adds a `$testthat` problem matcher (`contributes.problemMatchers`) that parses testthat 3.x progress-reporter failure headers — both `── Failure (file:line:col): name ──` and `── Error (…) ──` — so failing tests appear in the **Problems** panel with clickable file:line links.
- Paths resolve relative to `${workspaceFolder}/tests/testthat`, matching the directory testthat uses as its working directory while a test runs.
- Documents the matcher in `docs/r-package-dev.md` with a copy-pasteable `tasks.json` recipe.

Closes #206.

## Implementation notes

- Single-line pattern: captures `Failure|Error`, file, line, optional column, and the test name as the diagnostic message. Severity is fixed to `error`.
- Handles ASCII (`-- … --`) and box-drawing (`── … ──`) variants of the rule.
- Skip headers (`── Skipped (…) ──`) intentionally do not match — they're not problems.
- The matcher uses `severity: "error"` for both Failure and Error, matching the spec in #206. Promoting `Warning` headers can be added later if needed.

## Test plan

- [x] `cd editors/vscode && bun run compile:test` — TypeScript compiles
- [x] `bun test` (repo root) — all 406 tests pass, including the new 12 regex/contribution tests in `editors/vscode/src/test/problemMatchers.test.ts`
- [ ] Manual smoke test: open an R package in VS Code, add the documented `tasks.json` snippet, run the task, deliberately break a test, confirm the failure appears in the Problems panel with the correct file:line link

---

<a href="https://app.devin.ai/review/jbearak/raven/pull/217" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open in Devin Review
  </picture>
</a>

## Summary by CodeRabbit

* **New Features**
  * Added a `testthat` problem matcher for VS Code that parses R test failures/errors and shows results in the Problems panel with clickable file:line links.

* **Documentation**
  * Added a guide showing how to configure and run the matcher from a VS Code task, including path resolution details.

* **Tests**
  * Added unit tests validating the matcher across multiple reporter formats and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/jbearak/raven/pull/217)

### CODE REVIEW

File: editors/vscode/package.json [hunk 1]
```diff
@@ -360,6 +360,27 @@
...
+          "${workspaceFolder}/tests/testthat"
+        ],
+        "severity": "error",
+        "pattern": {
+          "regexp": "^[-─]+\\s+(Failure|Error)\\s+\\((.+?):(\\d+)(?::(\\d+))?\\):\\s+(.+?)\\s+[-─]+\\s*$",
+          "file": 2,
+          "line": 3,
+          "column": 4,
+          "message": 5
+        }
```
[the remark this answers is not in the document]
Comment (jbearak) [reply editors/vscode/package.json:381]: Pushing back. VS Code's `loop: true` does not concatenate lines into one problem — it creates a *new* problem per iteration, inheriting captures from earlier patterns. Per the [VS Code multi-line matcher docs](https://code.visualstudio.com/docs/debugtest/tasks#_defining-a-multi-line-problem-matcher):

> Multiple lines starting at the beginning of an output block can be matched by the same pattern, with each one creating a separate problem with the same file and message but different line/column information.

Applied to a real testthat failure:

```
── Failure (test-foo.R:12:3): process_data handles NAs ──
Expected: 1
Actual: 2
Backtrace:
   1. foo()
```

the suggested matcher would emit ~4 problems for one failure, all pointing at `test-foo.R:12:3`, with the test-name `message` overwritten on each iteration ("Expected: 1" / "Actual: 2" / …) — spamming the panel and burying the most useful piece of context.

The single-pattern shape produces one entry per failure with the test name as the message — universally informative whether the underlying failure is an `expect_equal` mismatch, a snapshot diff, or an uncaught error. The full expected/actual text stays in the terminal alongside the rest of the test output where the user is already reading it.

The two-pattern (no-`loop`) form sketched in #206 would instead overwrite `message` with whatever testthat happens to print on the very next line — informative for `expect_equal`, but misleading for snapshot failures (`Snapshot of code has changed:`) or rlang error headers (`Error in `f()`: …`). The single-pattern design here intentionally diverges from that sketch.

File: editors/vscode/src/test/problemMatchers.test.ts [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+    assert.ok(matcher, 'package.json must declare a problemMatcher named "testthat"');
+    return matcher;
+}
+
+function singlePattern(matcher: ProblemMatcherContribution): ProblemPattern {
+    assert.ok(
+        !Array.isArray(matcher.pattern),
+        'testthat matcher should declare a single-line pattern',
+    );
+    return matcher.pattern as ProblemPattern;
```
[the remark this answers is not in the document]
Comment (jbearak) [reply editors/vscode/src/test/problemMatchers.test.ts:85]: Linked to the matcher-shape decision in the package.json thread above — if the matcher stays single-pattern, this assertion is the right guard. It exists to catch an accidental switch to the multi-iteration shape that `loop: true` would impose, and to keep `singlePattern()` honest about returning a single `ProblemPattern`. We can relax it if a future change genuinely needs an array shape, but at that point we'd want a deliberate test covering whatever new semantics that brings.

**Comment by jbearak** [comment]
Verified the High finding and pushed a fix in 06e88a1. You were right: the original regex didn't match real output. Sampled testthat 3.3.2 + devtools 2.5.2 locally; the testthat source confirms the source-of-truth path is `expectation_location()` → `cli::format_inline("{.file {filename}:{srcref[1]}:{srcref[2]}}")`, which always single-quotes the location.

There are actually **three** live reporter formats, not one:

| Reporter | Sample header | When |
|---|---|---|
| `ProgressReporter` | `Failure ('test-sample.R:2:3'): expect_equal mismatch` | default for `devtools::test()` / `testthat::test_dir()` |
| `CompactProgressReporter` | `── Failure ('test-sample.R:2:3'): expect_equal mismatch ────` | opt-in; `cli::rule()`-wrapped form (also the ASCII fallback when `cli.unicode = FALSE`) |
| `LlmReporter` | `FAILURE: 'test-sample.R:2:3' ----------------------` | auto-selected by `testthat:::default_reporter()` when any of `CLAUDECODE` / `AGENT` / `GEMINI_CLI` / `CURSOR_AGENT` env vars is set |

The third explains what we were each seeing: both review bots and I were running under Claude Code, so testthat silently swapped in `LlmReporter`. End users in a plain VS Code terminal get `ProgressReporter` (format 1), which is what `devtools::test()` produces.

New regex anchors on the shared invariant — a quoted `'file:line:col'` — and handles all three shapes:

```
^[-─]*\s*(?:(Failure|Error)|(FAILURE|ERROR):)\s+\(?'([^':]+):(\d+)(?::(\d+))?'\)?(?::\s+(.+?))?\s*[-─]*\s*$
```

Capture-group indices shift to `file: 3, line: 4, column: 5, message: 6` to make room for the two mutually-exclusive label alternatives. `message` is `undefined` for `LlmReporter` since that reporter doesn't emit a test name — VS Code shows the file:line in that case.

Test fixtures now use captured real output for all three reporters (instead of hand-rolled samples), plus an explicit rejection of the prior unquoted shape so we can't regress to it. Empirically validated by piping `devtools::test()` output line-by-line through the regex with and without `CLAUDECODE` set: all 3 failures matched in both modes, with correct file/line/col/message captures. Bun test suite passes 406/406.
